### PR TITLE
feat(test): port fee swap and rebalance swap tests from Solidity reference

### PR DESCRIPTION
**Pull request by fgimenez**
ported from reference implementation https://github.com/tempoxyz/specs/blob/fee-manager/test/StableAMM.t.sol

ref #265

the tests are defined as unit tests bc the covered functionality is not exposed externally. also, some tests that hit overflow are ignored, same as in solidity

### CODE REVIEW

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs [hunk 1]
```diff
@@ -653,4 +654,362 @@ mod tests {
...
     }
+
+    fn setup_test_amm() -> (
+        TIPFeeAMM<'static, HashMapStorageProvider>,
+        Address,
+        Address,
+        Address,
+    ) {
+        let storage = Box::leak(Box::new(HashMapStorageProvider::new(1)));
+        let contract_address = Address::from([0x1u8; 20]);
```
Comment (0xKitsune) [inline crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: nit: very minor but we can use any address here. No change needed but you can do `TIPFeeAMM::new(Address::ZERO storage)` if you prefer to reduce loc.

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs [hunk 2]
```diff
@@ -653,4 +654,362 @@ mod tests {
...
+    }
+
+    /// Test basic fee swap functionality
+    /// Corresponds to testFeeSwap in StableAMM.t.sol
+    #[test]
+    fn test_fee_swap() {
+        let (mut amm, _, user_token, validator_token) = setup_test_amm();
+
+        // Setup pool with 100,000 tokens each
+        let liquidity_amount = uint!(100000_U256) * uint!(10_U256).pow(U256::from(18));
```
Comment (0xKitsune) [inline crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: `TIP20` tokens conform to specific currency decimals. The only decimals supported at the moment is USD with 6 decimals. I think this should be fine for tests though.

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs [hunk 3]
```diff
@@ -653,4 +654,362 @@ mod tests {
...
+
+        // Execute fee swap for 1000 tokens
+        let amount_in = uint!(1000_U256) * uint!(10_U256).pow(U256::from(18));
+
+        // Calculate expected output: amountIn * 0.9975
+        let expected_out = (amount_in * M) / SCALE;
+
+        // Execute fee swap
+        let result = amm.fee_swap(user_token, validator_token, amount_in);
+        assert!(result.is_ok(), "Fee swap should succeed");
```
Comment (0xKitsune) [inline crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: Since we are only returning `TIPFeeAMMError`, we can update the test to return `Result<(), TIPFeeAMMError>` and propagate this.

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs [hunk 4]
```diff
@@ -653,4 +654,362 @@ mod tests {
...
+
+        // Execute fee swap - should fail
+        let result = amm.fee_swap(user_token, validator_token, too_large_amount);
+
+        match result {
+            Err(TIPFeeAMMError::InsufficientLiquidity(_)) => {
+                // Expected error
+            }
+            _ => panic!("Expected InsufficientLiquidity error"),
+        }
```
Comment (0xKitsune) [inline crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: We can use `assert!(matches!(result, TipFeeAMMError::insufficient_liquidity()))` here.

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs [hunk 5]
```diff
@@ -653,4 +654,362 @@ mod tests {
...
+        );
+
+        // Execute multiple fee swaps
+        let swap1 = uint!(1000_U256) * uint!(10_U256).pow(U256::from(18));
+        let swap2 = uint!(2000_U256) * uint!(10_U256).pow(U256::from(18));
+        let swap3 = uint!(3000_U256) * uint!(10_U256).pow(U256::from(18));
+
+        assert!(amm.fee_swap(user_token, validator_token, swap1).is_ok());
+        assert!(amm.fee_swap(user_token, validator_token, swap2).is_ok());
+        assert!(amm.fee_swap(user_token, validator_token, swap3).is_ok());
```
Comment (0xKitsune) [inline crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: Same thing here, if we update the test to return `Result<(), TIPFeeAMMError>` we can propagate these instead.

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs [hunk 6]
```diff
@@ -653,4 +654,362 @@ mod tests {
...
+        let imbalance_before = if x > y { x - y } else { y - x };
+        let imbalance_after = if new_x > new_y {
+            new_x - new_y
+        } else {
+            new_y - new_x
+        };
+        assert!(
+            imbalance_after < imbalance_before,
+            "Swap should reduce imbalance"
+        );
```
Comment (0xKitsune) [inline crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: If we are testing rebalance swap logic can we invoke `execute_rebalance_swap()` on the pool?

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs [hunk 7]
```diff
@@ -653,4 +654,362 @@ mod tests {
...
+        // For a balanced pool, swapping user tokens would mean:
+        // new_x = x + swap_amount (increasing user tokens)
+        // This is wrong direction since pool is already balanced
+
+        // In the actual implementation, this check happens:
+        // if new_x >= x { return Err(InvalidSwapCalculation) }
+
+        // Since we're swapping FROM validator TO user in rebalance_swap,
+        // and pool is balanced, this would unbalance it
+        assert_eq!(x, y, "Pool should be balanced initially");
```
Comment (0xKitsune) [inline crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: Same thing here if we are testing rebalancing in the wrong direction can we invoke `execute_rebalance_swap()`? Looks like this test might not be complete?

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs [hunk 8]
```diff
@@ -653,4 +654,362 @@ mod tests {
...
+        // Calculate new x
+        let result = amm.calculate_new_reserve(known_y, l);
+        assert!(result.is_ok(), "Should calculate new reserve");
+
+        let new_x = result.unwrap();
+        assert!(new_x > 0, "New reserve should be positive");
+
+        // Test error handling with extreme values
+        let large_y = uint!(1_000_000_U256) * uint!(10_U256).pow(U256::from(6));
+        let _zero_result = amm.calculate_new_reserve(large_y, l);
```
Comment (0xKitsune) [inline crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: Should there be some assertion against this value?

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs (hunk 1 above)
Comment (fgimenez) [reply crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: makes sense, done

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs (hunk 2 above)
Comment (fgimenez) [reply crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: good catch, updated to use 6 decimals

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs (hunk 3 above)
Comment (fgimenez) [reply crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: ok, done

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs (hunk 4 above)
Comment (fgimenez) [reply crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: sure done

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs (hunk 5 above)
Comment (fgimenez) [reply crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: yep done

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs (hunk 6 above)
Comment (fgimenez) [reply crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: yeah much better, done

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs (hunk 7 above)
Comment (fgimenez) [reply crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: indeed good point, fixed now

File: crates/precompiles/src/contracts/tip_fee_manager/amm.rs (hunk 8 above)
Comment (fgimenez) [reply crates/precompiles/src/contracts/tip_fee_manager/amm.rs]: right, this test is currently ignored but added the assertion in any case
